### PR TITLE
Improve RDMA rendezvous speed

### LINKED ISSUE #10530: Protobuf serialization slows down RDMA
According to a test similar to https://github.com/tensorflow/tensorflow/issues/6116, the current RDMA implementation has performance issues.

This tensor copy throughput according to [this test](https://gist.github.com/yaroslavvb/1124bb02a9fd4abce3d86caf2f950cb2) (change `assign_add` to `assign` in L53, and test with tensor size 100MB):
```
Distributed rate: 536.72 MB per second
```
Which is just slightly higher than the gRPC implementation (the gRPC fixed version: https://github.com/tensorflow/tensorflow/pull/7466)

Here is the profiling result:
[device1 worker profiling report](https://github.com/tensorflow/tensorflow/files/1060989/dev1.pdf)
[device2 worker profiling report](https://github.com/tensorflow/tensorflow/files/1060990/dev2.pdf)

There are some obvious issue, for example, the [string resize issue](https://github.com/google/protobuf/blob/master/src/google/protobuf/stubs/stl_util.h#L67), the unnecessary memory copy and the single threaded serialization/deserialization.

### PR DESCRIPTION
**Pull request by llhe**
This PR fixes https://github.com/tensorflow/tensorflow/issues/10530.

Tested with RoCE setting: Mellanox Technologies MT27620 Family

Before this PR:
```
Distributed rate: 536.72 MB per second
```
After this PR:
```
Distributed rate: 1346.06 MB per second
```

TODO: memcpy is still slow and need to improved.

### CODE REVIEW

**Comment by byronyi** [comment]
Seems a reasonable patch. 

By the way, how do you measure the rates? The MT27620 ConnectX-4 NIC should support 100 GbE fabrics. These rates seem a little bit too small compared to the hardware capabilities.

**Comment by llhe** [comment]
It's not pure bandwidth test but a tensor copy test like this: https://gist.github.com/yaroslavvb/1124bb02a9fd4abce3d86caf2f950cb2. (change `assign_add` to `assign` in L53).

**Comment by byronyi** [comment]
Thanks for pointing me to the script. I am actually wondering if the loopback RDMA performance could reflect the actual in-fabric performance. Honestly I am not sure if hardware is involved at all when using loopback; it might use a completely different mechanism than NIC doing DMA.

Nevertheless this patch will still help in the case of local worker-ps communication.

**Comment by llhe** [comment]
@junshi15 @anfeng @shamoya  Any comments on this?

**Comment by 2sin18** [comment]
It doesn't converge in GPU device if I patch this PR in tensorflow master branch. It's OK in CPU device, so I guess there're some bugs in `SetProtoFromGPUSync` related code.

**Comment by byronyi** [comment]
@2sin18 Note that the original implementation uses ``SetProtoFromGPUSync`` even before this patch. If there is anything susceptible it should be ``CopyGPUTensorToCPUSync`` and ``CopyCPUTensorToGPUSync `` which is introduced in this patch.

File: tensorflow/contrib/verbs/rdma_rendezvous_mgr.cc [hunk 1]
```diff
@@ -99,12 +100,41 @@ void RdmaRemoteRendezvous::RecvFromRemoteAsync(
...
+          Allocator* alloc = ProcessState::singleton()->GetCUDAHostAllocator(0);
+          Tensor copy(alloc, rm.data_type_, rm.tensor_shape_);
+          memcpy(DMAHelper::base(&copy), input, rm.tensor_bytes_);
+
+          Allocator* dst_alloc = dst_dev->GetAllocator(host_alloc_attrs);
+          Tensor gpu_copy(dst_alloc, rm.data_type_, rm.tensor_shape_);
+          s = VerbsUtil::CopyCPUTensorToGPUSync(&copy, recv_args.device_context,
+                                                dst_dev, &gpu_copy);
+          CHECK(s.ok()) << "copy tensor to gpu sync";
+          val = std::move(gpu_copy);
```
Comment (junshi15) [inline tensorflow/contrib/verbs/rdma_rendezvous_mgr.cc]: Does std::move work properly for GPUTensor?

Comment (junshi15) [inline tensorflow/contrib/verbs/rdma_rendezvous_mgr.cc]: The chain of copying can be further simplified. Currently, you copy twice here (assuming move works as expected):
input --> (cpu) copy --> gpu_copy -->(move)--> val,
If you want to be aggressive, you can do input-->gnu_copy directly, skipping the cpu_copy. In one of our early versions (not released), we had that. But it required low-level functions similar to cudaMemcpy. It was too complex, so in the final public version, we chose the protobuf route for simplicity, but sacrificed performance. Ditto on the sender side (rdma.cc).

Comment (junshi15) [inline tensorflow/contrib/verbs/rdma_rendezvous_mgr.cc]: @llhe Thanks for your contribution. The patch looks good except that I am not sure whether std::move works for GPU tensors as expected. Besides speed, can you verify the result is correct? @2sin18 had issues about the convergence while running with GPU. I suspect it has something to do either "copy" or "move" operations.

**Comment by llhe** [comment]
@2sin18 Did you tested with RDMA? Can you provide a repro about the convergence issue? Thanks.

File: tensorflow/contrib/verbs/rdma_rendezvous_mgr.cc (hunk 1 above)
Comment (llhe) [inline tensorflow/contrib/verbs/rdma_rendezvous_mgr.cc]: std::move should be ok for GPU tensor, see the code [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/tensor.h#L692). Actually the [standard assignment](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/tensor.h#L161) also just an pointer assignment.

File: tensorflow/contrib/verbs/rdma_rendezvous_mgr.cc [hunk 2]
```diff
@@ -99,12 +100,41 @@ void RdmaRemoteRendezvous::RecvFromRemoteAsync(
...
+        if (dst_dev->tensorflow_gpu_device_info() &&
+            (!recv_args.alloc_attrs.on_host())) {
+          CHECK(recv_args.device_context)
+            << "send dev name: " << src_dev->name()
+            << " gpu_info: " << src_dev->tensorflow_gpu_device_info();
+          Allocator* alloc = ProcessState::singleton()->GetCUDAHostAllocator(0);
+          Tensor copy(alloc, rm.data_type_, rm.tensor_shape_);
+          memcpy(DMAHelper::base(&copy), input, rm.tensor_bytes_);
+
+          Allocator* dst_alloc = dst_dev->GetAllocator(host_alloc_attrs);
```
Comment (llhe) [inline tensorflow/contrib/verbs/rdma_rendezvous_mgr.cc]: @2sin18 @junshi15 Thanks for pointing this out. This should be changed to `recv_args.alloc_attrs`?

Comment (junshi15) [inline tensorflow/contrib/verbs/rdma_rendezvous_mgr.cc]: yes, I think "recv_args.alloc_attrs" is the right attributes.

**Comment by junshi15** [comment]
@2sin18 Does @llhe 's new patch fix the convergence problem?

**Comment by 2sin18** [comment]
@junshi15 Thanks to @llhe 's patch, it works now.

File: tensorflow/contrib/verbs/rdma.cc [hunk 3]
```diff
@@ -703,21 +703,47 @@ void RdmaTensorBuffer::SendNextItem() {
...
-        // "val" is on a GPU. Uses GPUUtil to fill the proto.
-        s = VerbsUtil::SetProtoFromGPUSync(
-            in, src_dev, send_args.device_context, &proto, is_dead);
-        CHECK(s.ok()) << "set proto from gpu sync";
+          << "send dev name: " << src_dev->name()
+          << " gpu_info: " << src_dev->tensorflow_gpu_device_info();
+
+        if (can_memcpy) {
+          AllocatorAttributes host_alloc_attrs;
+          host_alloc_attrs.set_gpu_compatible(true);
```
Comment (poxvoculi) [inline tensorflow/contrib/verbs/rdma.cc]: Add
host_alloc_attrs.set_nic_compatible(true);
This indicates that the memory should be allocated from a pool registered for DMA by the NIC.   Not effective yet, but probably soon.

**Comment by poxvoculi** [comment]
Jenkins test this please

**Review by poxvoculi — APPROVED** [review]

**Pull request merged**